### PR TITLE
Feature/179+184 enemy ai

### PR DESCRIPTION
**Pull request by TerMyY**
[Task (Movement)](https://s24109.atlassian.net/browse/SCRUM-179) [Task (Combat)](https://s24109.atlassian.net/browse/SCRUM-184)

### Current state

Still very much a work in progress, but for the sake of testing and continuing further work from content side, shipping sort of working prototype now. Some features are missing, some are broken, many things need refactoring and readjusting.

For now however we at least will have the prefabs that we can work with, and some demonstrable enemy behavior.

### Known bugs (some of them)

-In some places enemies will recognize solid walls as a cliff. Likely an issue with level terrain colliders.

-When flipping the direction, enemies will displace their collision and sometimes fall/clip due to that. Probably an issue with scaling. (FIXED)

-Ranged attack is just broken completely, I wrote it in haste, will refactor.

-While being distracted and staying near the place of distraction, enemies will damage the player. No idea why, I'll investigate.

### Why is it in a single giant file

I've tried to separate the behavior into modules, but along the way most things required a lot of duplication or good planning which I lack because I was just throwing code at the wall and seeing what sticks. So rn it is bundled in a single (very confusing) file and all the enemies use the same script with adjusted parameters. Btw this script was written from scratch again compared to the old one, and it ended up as an even bigger mess anyway.

### CODE REVIEW

File: Assets/Scripts/EnemyAI.cs [hunk 1]
```diff
@@ -1,204 +1,287 @@
...
+                {
+                    state = State.Chase;
+                    break;
+                }
+
+                if (_distanceToTarget < seeRange && direction == _targetDirection)
+                {
+                    state = State.Chase;
+                    break;
+                }
```
Comment (FrozzLab) [inline Assets/Scripts/EnemyAI.cs:144]: Add check for obstacles in between enemy and player

File: Assets/Scripts/EnemyAI.cs [hunk 2]
```diff
@@ -1,204 +1,287 @@
...
-            if (_state == State.Return)
+            case State.Patrol:
             {
-                Patrol();
-            }
-        }
+                _speed = walkSpeed;
+                
+                var isGroundAhead = Physics2D.Raycast(new Vector2(_colliderBoundsReference.center.x + (int)direction, _colliderBoundsReference.center.y), Vector2.down, 1.5f, _groundLayer);
+                Debug.DrawRay(new Vector2(_colliderBoundsReference.center.x + (int)direction, _colliderBoundsReference.center.y), Vector2.down, isGroundAhead ? Color.green : Color.red);
```
Comment (FrozzLab) [inline Assets/Scripts/EnemyAI.cs:117]: add "flags" to limit patrol range

**Review by FrozzLab — APPROVED** [review]
Apart from the known issues all seems fine, Approved but fix in separate PR.

**Review by WildCup — CHANGES_REQUESTED** [review]
make the collider center of the object. It's not logical to offset it. Like think what do you understand by "enemy position" right now it's an empty space a few meters right to the enemy.
If you make a child object 'Sprite' you can offset the sprite anyway you want. The the problem with flipping is gone.

image
image

File: Assets/Scripts/EnemyAI.cs [hunk 3]
```diff
@@ -1,204 +1,289 @@
...
     }
 
- 	public void NoticeDistraction(Vector3 position)
-	{
-		var magnitude = (position - transform.position).magnitude;
-		if(magnitude > detectRange) return;
+    public void SetTarget(GameObject newTarget, bool isLoud, int priority)
+    {
+        target = newTarget;
+        isTargetLoud = isLoud;
```
Comment (WildCup) [inline Assets/Scripts/EnemyAI.cs:236]: what is isTargetLoud for?

File: Assets/Scripts/EnemyAI.cs [hunk 4]
```diff
@@ -1,204 +1,289 @@
...
         {
-            _speed = 0;
-            return;
+            Jump();
         }
-
-        _speed = chaseSpeed;
+        
+        if (_distanceToTargetHorizontal < attackRange && state == State.Chase) return;
+        _rigidbody.velocity = new Vector2((int)direction * _speed * Time.deltaTime, _rigidbody.velocity.y);
```
Comment (WildCup) [inline Assets/Scripts/EnemyAI.cs:199]: small thing but in FixedUpdate you should use Time.fixedDeltaTime

**Review by WildCup — APPROVED** [review]
approving as this is blocking other' developers progress

**Pull request merged**